[ConstExpr] Emit traceback when failing during nested call evaluation

jmolloy opened the pull request:
When emitting a constexpr-evaluation failure message, this change emits
a trace of the calls the evaluator followed to generate the failure.

To make the output easier to read, calls on the same source line as the
failing location are elided from the output. So for example:

func foo() -> Bool {
  print("Un-constant")
  return true
}

test.swift:6:1: error: #assert condition not constant
^
test.swift:2:9: note: expression not evaluable as constant here
  print("Un-constant")

Notice there is no stack trace for the evaluation of the lambda.

Commit:
[ConstExpr] Emit traceback when failing during nested call evaluation

When emitting a constexpr-evaluation failure message, this change emits
a trace of the calls the evaluator followed to generate the failure.

To make the output easier to read, calls on the same source line as the
failing location are elided from the output. So for example:

func foo() -> Bool {
  print("Un-constant")
  return true
}

test.swift:6:1: error: #assert condition not constant
^
test.swift:2:9: note: expression not evaluable as constant here
  print("Un-constant")

Notice there is no stack trace for the evaluation of the lambda.

jmolloy commented:
@swift-ci please test tensorflow

mhong commented on lib/SIL/SILConstants.cpp:
```diff
@@ -479,20 +496,27 @@ void SymbolicValue::emitUnknownDiagnosticNotes(SILLocation fallbackLoc) {
+    if (SM.getLineNumber(sourceLoc) == originalDiagnosticLineNumber)
+      continue;
+
+    auto diag = emittedFirstNote ? diag::constexpr_called_from
```
also consider adding a test case where diag::constexpr_called_from is exercised 2 or more times?

lattner commented on include/swift/SIL/SILConstants.h:
```diff
@@ -147,15 +146,15 @@ class SymbolicValue {
     UnknownReason unknown_reason;
 
-    //unsigned integer_bitwidth;
+    llvm::SmallVector<SourceLoc, 4> callStack;
```
SymbolicValue is designed to be a POD type that can be thrown around without running its destructor, and sizeof(SymbolicValue) needs to remain small (lots of them get put in densemaps etc).

Please make the callstack be a new tail allocated node, similar to how the apint represent or the aggregate value representation is stored.  We can then have two "unknown" representations, one that has a stack trace (which is seperately allocated and one with an inline representation).

lattner commented on include/swift/SIL/SILConstants.h:
```diff
@@ -196,24 +195,30 @@ class SymbolicValue {
   std::pair<SILNode *, UnknownReason> getUnknownValue() const {
     assert(representationKind == RK_Unknown);
-    return { value.unknown, aux.unknown_reason };
+    return {value.unknown, aux.unknown_reason};
```
Please split formatting changes out from behavior changing patches :-)

jmolloy replied:
Way ahead of you, my git-fu has rusted over the past year though :) Thanks for the comments, will address soon.

Commit:
Clang-format touched files to make diff smaller.

jmolloy commented:
Thanks both. The patch is now split into clang-format only and functional changes. I will send another PR to add a clang-format base style file and format all files we've introduced on this branch.

We now use the BumpPtrAllocator as Chris suggested. I've created a new struct UnknownInfo which is hungoff and contains the SILNode, reason and call stack.

PTAL.

jmolloy commented:
@swift-ci please test tensorflow

lattner commented on include/swift/SIL/SILConstants.h:
```diff
@@ -146,14 +148,20 @@ class SymbolicValue {
+    UnknownReason reason;
+
+    /// The call stack while evaluating the unknown value.
+    llvm::SmallVector<SourceLoc, 4> call_stack;
```
The objects associated with SymbolicValue are allocated in the bump pointer allocator ... so their destructors are never called.  This means they cannot contain non-pod values like smallvector (it would leak memory when it contained values larger than its "small" size of 4).  Please switch this to use the same pattern as the AggregateValue.  You should use a trailing object allocation of the SourceLoc objects.  This will make the whole thing POD and in a continuous allocation.

lattner replied:
Also, please declare that class in the .cpp file, so the public interface of SymbolicValue is kept small.

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -769,21 +781,24 @@ ConstExprFunctionState::computeOpaqueCallResult(ApplyInst *apply,
     // its string and file/line/col info and propagate it up!
-    return SymbolicValue::getUnknown((SILInstruction *)apply,
-                                     UnknownReason::Trap);
+    return SymbolicValue::getUnknown(
```
I'd suggest adding a "getUnknown" method on the enclosing class that takes a node + reason, but handle passing the callStack and allocator.

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.h:
```diff
@@ -43,6 +44,9 @@ class ConstExprEvaluator {
   llvm::BumpPtrAllocator allocator;
 
+  /// The current call stack, used for providing accurate diagnostics.
+  llvm::SmallVector<SourceLoc, 4> callStack;
```
This smallvector is just fine!

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.h:
```diff
@@ -52,6 +56,15 @@ class ConstExprEvaluator {
 
   llvm::BumpPtrAllocator &getAllocator() { return allocator; }
 
+  void pushCallStack(SourceLoc loc) { callStack.push_back(std::move(loc)); }
```
SourceLoc is a single word POD type, no need to move it.

lattner commented:
@jmolloy one other random thing.  It would be great if you could split the clang-format change out to a separate PR as well.  The standard process for swift is "merge" commits, which make it hard to use blame afterward if a big formatting change is included into behavior changing diffs.
